### ISSUE TITLE
Slow and bad

### ISSUE THREAD

**Issue by mjacobsen32** [mjacobsen32/ramblingrealms-trading#64]
My DRL algorithm has a really slow train process, around 150 iterations per second, and that is too slow for me. The reason it is slow is because I maintain things such as the holdings per symbol, the open positions, etc... so I can enforce trade rules such as not going into negative available cash, being rewarded for returns against the portfolio value, etc... the house keeping is what is slowing things down, but I think that should also make it more accurate... how can I speed things up? Part of me wants to remove all the house keeping stuff, and just have a signal to the agent whether the purchase was good or bad idea based on local maximum and minimums, what are some other ideas for speeding up training and maybe getting a better reward signal? Right now I need to maintain a running returns list for each symbol so it can calculate the sharpe ratio each step... but I feel like these things on a per step basis slow it down too much. Maybe I should only be calculating these things after it runs through the entire train set? so I would let it run wild through the entire training set, and then the end of that I would signal if it was good or bad based on profits, but then each "step" is actually the entire data set, and that's probably not a very good signal...

Your goal is to drastically speed up the algorithm up to 10,000 iterations per second, and maintain a constant time internal state in the training steps such that the iterations per second is maintained at 10,000. This requires constant state and no longer maintaining any historical information.

Feel free to take creative liberty to transform things as much as you would like, the unit tests and tests can completely change with the new api you may or may not create. However, there still needs to be a testing mode which will maintain the internal state such that it can enforce internal trading constraints such as non-negative cash flow, etc...

Another note, it may be best to refactor to create a single environment base class, and have a derived training class, and a derived trading / testing class, such that the derived testing and trading class will maintain positions, cash, etc... and keep track of all of the internal state required for realistic tests and trades, but the training environment's goal is to be fast and as stateless as possible to prevent the iterations from being a function of n, feel free to run wild with the change, I may or may not keep them but am curious what you are able to cook up. 

All of the existing functionality from the command line must remain relatively unchanged, the alg train, backtest, and paper-trade need to continue to do what they do. The main overhaul here is to simplify the training step loop and inherit.

You are also able to modify any hyperparameters or tune any of the agent and environment parameters to make it faster, more accurate, and easier to observe. Hyper parameter tuning would be a great achievement. I would also value a more fine tuned testing library which will be able to run the train and backtest and get real results, results that make sense and prove that the model is learning to maximize profit returns.
